gateway/ln-gateway/src/lib.rs

```diff
@@ -511,112 +511,117 @@ impl Gateway {
     }
 
     pub async fn run(mut self, listen: SocketAddr, password: String) -> Result<()> {
-        let mut tg = self.task_group.clone();
-
         let sender = GatewayRpcSender::new(self.sender.clone());
-        tg.spawn("Gateway Webserver", move |server_ctrl| async move {
-            let mut webserver = tokio::spawn(run_webserver(password, listen, sender));
-
-            // Shut down webserver if requested
-            if server_ctrl.is_shutting_down() {
-                webserver.abort();
-                let _ = futures::executor::block_on(&mut webserver);
-            }
-        })
-        .await;
+        self.task_group
+            .spawn("Gateway Webserver", move |_| async move {
+                run_webserver(password, listen, sender)
+                    .await
+                    .expect("Error starting webserver");
+            })
+            .await;
 
         // TODO: try to drive forward outgoing and incoming payments that were
         // interrupted
-        let loop_ctrl = tg.make_handle();
-        loop {
+        let loop_ctrl = self.task_group.make_handle();
+        let shutdown_sender = self.sender.clone();
+        loop_ctrl
+            .on_shutdown(Box::new(move || {
+                Box::pin(async move {
+                    // ignore errors: we don't care if already stopped
+                    let _ = shutdown_sender.send(GatewayRequest::Shutdown()).await;
+                })
+            }))
+            .await;
+
+        // Handle messages from webserver and plugin
+        while let Some(msg) = self.receiver.recv().await {
```

Comment:
Seems to me like we should just relaying on the sender getting dropped would be enough, but I guess this way works as well. I'm not sure which way is better.

If a task is just an async handler for another thread, just relaying on channels being disconnected is the cleanest, IMO. And it doesn't matter if the other side is stuck or whatever, as it will not send new messages if it's stuck.

Reply:
Tested 2feb9610bf7c7f99a6a1a67a86ecc537339da3b1 multiple times when debugging the CI pipeline and had no issues.

Reply:
I tried doing that and was running into issues causing the tests to timeout. What ended up working is implementing `on_shutdown` and sending a signal to shutdown both the event loop and the webserver.